Core/Script Implement Atonement discipline priest.

HannibalRoG opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

-  Implement Talent Atonement for discipline priest
-  Modify Power Word: Shield for cast Atonement proc

Be gentle...

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [x] master

**Issues addressed:** Closes #  (insert issue tracker number)
There are none

**Tests performed:** (Does it build, tested in-game, etc.)
Compiled, tested in game

**Known issues and TODO list:** (add/remove lines as needed)

- [ ] Implement other spell affected by Atonement
- [x] When you use Power word: shield without finishing the previous one does not reapply the aura Atonement on target.


<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

HannibalRoG committed:
Implement Atonement discipline priest.
Now only with power word: shield

Todo:

Implement others spells in .cpp

- Plea
- Shadow mend
- Power word: radiance

HannibalRoG committed:
Merge branch 'master' into Atonement

HannibalRoG committed:
Solve the issue that cause not reapply aura Atonement with PWS active.

Shauren commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -164,6 +167,77 @@ class spell_pri_aq_3p_bonus : public SpellScriptLoader
+            });
+
+            if (playerGroupList.size() > 1)
+                playerGroupList.sort(Trinity::HealthPctOrderPred());
```
What is the point of sorting if you cast SPELL_PRIEST_ATONEMENT_HEAL on all units in this list?

Shauren commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -164,6 +167,77 @@ class spell_pri_aq_3p_bonus : public SpellScriptLoader
+            Player* player = GetCaster()->ToPlayer();
+
+            SpellInfo const* spellInfoAtonement = sSpellMgr->GetSpellInfo(SPELL_PRIEST_ATONEMENT_AURA);
+            if (!spellInfoAtonement || !player->HasAura(SPELL_PRIEST_ATONEMENT_AURA))
```
This condition can never fail. SPELL_PRIEST_ATONEMENT_AURA always exists in this hook (!spellInfoAtonement) and its the aura being scripted (HasAura)

Shauren replied:
This condition can never fail. SPELL_PRIEST_ATONEMENT_AURA always exists in this hook (!spellInfoAtonement) and its the aura being scripted (HasAura)

Shauren commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -164,6 +167,77 @@ class spell_pri_aq_3p_bonus : public SpellScriptLoader
+            if (!spellInfoAtonement || !player->HasAura(SPELL_PRIEST_ATONEMENT_AURA))
+                return;
+
+            if (player->GetUInt32Value(PLAYER_FIELD_CURRENT_SPEC_ID) != TALENT_SPEC_PRIEST_DISCIPLINE)
```
Don't check this - only disc priests should have this in SpecializationSpells.db2

Shauren commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -164,6 +167,77 @@ class spell_pri_aq_3p_bonus : public SpellScriptLoader
+            DamageInfo* damageInfo = eventInfo.GetDamageInfo();
+            int32 heal = CalculatePct(damageInfo->GetDamage(), spellInfoAtonement->GetEffect(EFFECT_0)->CalcValue());
+
+            for (auto itr : playerGroupList)
```
Codestyle/naming - in loops using `for (var : container)` syntax the var is not an iterator (dont name it itr) and don't use auto (because its not an iterator)

tkrokli commented on sql/updates/world/master/RoG_2017_06_13_00_world.sql:
```diff
@@ -0,0 +1,8 @@
+	Spell ID:		81749
+	Spell Name:		Atonement
+	Spell Family: 	SPELLFAMILY_PRIEST
+*/
```
3 changes needed in this file:
1. add NewLine at End of File (currently missing)
2. never use apostrophes around integer number values (spell_id)
3. TC codestyle: use single line `--` instead of `/* */` for short comments. ( `/* */` is for GPL headers and long explanations in big files.)

HannibalRoG committed:
Fix all reviews from Shauren, thanks.

HannibalRoG committed:
Fix .sql with code standard, thanks tkrokli.

HannibalRoG commented:
@shauren new review please.

wonder2011 commented:
Tested in game. Working nice.

Commit:
Update SQL

- remove redundant blank lines before and after comment
- add required space between comment marker and text

Keader commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -164,6 +167,69 @@ class spell_pri_aq_3p_bonus : public SpellScriptLoader
+
+        void HandleProc(AuraEffect const* aurEff, ProcEventInfo& eventInfo)
+        {
+            Player* player = GetCaster()->ToPlayer();
```
This can be null, no ?!

wonder2011 replied:
@Keader i thing not because up on CheckProc is verified this, or i miss something ?

mdX7 replied:
target may be null in AuraScripts - GetCaster() can not be null, because it is the Unit causing the proc.

Keader replied:
Yes, will check if GetCaster() is not null, but are you sure that is player ?

wonder2011 replied:
@Keader well now i see your point. @HannibalRoG i recommend that you validate that caster is player, something like

```
if (Player* player = GetCaster()->ToPlayer())
{
    DoAllThingsHere();
}
```
Right @Keader ??

Keader replied:
y

HannibalRoG replied:
ok, i see the point.

HannibalRoG committed:
Fix for Keader recomendations.

HannibalRoG committed:
Merge branch 'Atonement' of https://github.com/HannibalRoG/TrinityCore into Atonement

Shauren commented:
I looked at this script and well, its wrong - atonement on retail works on all targets affected by it, not just group members

Shauren commented:
Also from ingame testing - this script ignores mastery

Shauren commented:
Replaced by 6c9b5e4cf59ef94b1ad3e32c435de430b75fe811

wonder2011 commented:
@Shauren but you implement some new fratures in the Core when you see this pull request right?
I thing just now, this spells will be hard to implement with the curren SpellScript system:

```
186263
187464
186439
```

Or i'm wrong ?

Shauren commented:
No, they won't be hard

wonder2011 commented:
@Shauren then i need to learn much more, tks !
